pennylane/vqe/vqe.py

```diff
@@ -174,35 +191,39 @@ def simplify(self):
           (-1) [X0]
         + (1) [Y2]
         """
-        coeffs = []
+        data = []
         ops = []
 
-        for c, op in zip(self.coeffs, self.ops):
+        for i in range(len(self.ops)):  # pylint: disable=consider-using-enumerate
+            op = self.ops[i]
+            c = self.coeffs[i]
             op = op if isinstance(op, Tensor) else Tensor(op)
 
             ind = None
-            for i, other in enumerate(ops):
+            for j, other in enumerate(ops):
                 if op.compare(other):
-                    ind = i
+                    ind = j
                     break
 
             if ind is not None:
-                coeffs[ind] += c
-                if np.allclose([coeffs[ind]], [0]):
-                    del coeffs[ind]
+                data[ind] += c
+                if np.isclose(qml.math.toarray(data[ind]), np.array(0.0)):
```

Comment:
Not sure if there is a better way? :) `isclose` compares to 0.0 too, no?

Reply:
In a bunch of the transforms I wrote, I ended up having to do something like the below to work in all interfaces:
```python
# U is a tensor in some interface 
if qml.math.allclose(U[0, 1], [0.0]):

# angles is a 3-element tensor in some interface
if not qml.math.allclose(angles, qml.math.zeros(3), atol=atol, rtol=0):
```
This avoids having to convert stuff to numpy.

Reply:
I fear that ``math.allclose`` _does_ convert stuff to numpy as it is ;)

Reply:
Yep :( Currently, `math.allclose` does the following:
```python
t1 = ar.to_numpy(a)
t2 = ar.to_numpy(b)
res = np.allclose(t1, t2, rtol=rtol, atol=atol, **kwargs)
```

While this works in most cases, it _does not work when using a JIT_. E.g., if you decorate your QNode/cost function with `@jax.jit` or `@tf.function`, then it will complain at this line.

This is a restriction of using the JIT --- on first pass, the cost function is traced with _valueless_ tensors, compiled, and only later are the actual values passed to the compiled function. So we can never convert to NumPy arrays inside the jit.

A better solution would be to update `qml.math.allclose` to stay autodiff agnostic:

```python

def allclose(a, b, rtol=1e-05, atol=1e-08, **kwargs):
    # get the interface
    interface = _multi_dispatch([a, b])

    # need to cast all values to the same type
    a, b, rtol, atol = qml.math.coerce([a, b, rtol, atol], like=interface)

    return qml.math.all(qml.math.abs(a - b) <= (atol + rtol * qml.math.abs(b)))
```

Reply:
So nice that you already provide the solution ;)

Reply:
Shall I open an issue for this? Or a ticket?

Reply:
Maybe an issue? I'm not sure if my code solution above is fully working or not 😆

Reply:
> I fear that math.allclose does convert stuff to numpy as it is ;)

Oh :( 

> qml.math.coerce([a, b, rtol, atol], like=interface)

Wait is `coerce` a real thing? That's super helpful...